File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys
sys.setrecursionlimit(10 ** 6)

input = sys.stdin.readline
n = int(input())
area = [list(map(int, input().split())) for _ in range(n)]
max_height = max(map(max, area))
max_cnt = 0
dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

def is_valid_coord(y, x):
    return 0 <= y < n and 0 <= x < n

def dfs(y, x):
    chk[y][x] = 1
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        if (is_valid_coord(ny, nx) and not chk[ny][nx] and area[ny][nx] >= height):
            dfs(ny, nx)

for height in range(0, 100): 
    cnt = 0
    chk = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if area[i][j] >= height and not chk[i][j]:
                cnt += 1
                dfs(i, j)

    if cnt > max_cnt: max_cnt = cnt

print(max_cnt)

File: 백준/Silver/2468. 안전 영역/안전 영역.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let inp = fs.readFileSync(filePath).toString().split("\n");
const n = Number(inp.shift());
const area = []; // inp.map(el => el.split(' ').map(Number));
let maxHeight = 1;
let height;
inp.forEach(el => {
    area.push(el.split(' ').map(Number));
    if (Math.max(...area[area.length - 1]) > maxHeight) 
        maxHeight = Math.max(...area[area.length - 1]); 
})
let chk;
let cnt = 0;
let maxCnt = 0;
let dy = [-1, 0, 1, 0];
let dx = [0, 1, 0, -1];


const isValidCoord = (y, x) => 0 <= y && y < n && 0 <= x && x < n;

function dfs(y, x) {
    chk[y][x] = 1;
    for (let i = 0; i < 4; i++) {
        const ny = y + dy[i];
        const nx = x + dx[i];
        if (isValidCoord(ny, nx) && area[ny][nx] >= height && !chk[ny][nx]) {
            dfs(ny, nx);
        }
    }
}

function findMaxHeight() {
    for (height = 0; height < 100; height++) {
        chk = Array.from(Array(n), () => Array(n).fill(0));
        cnt = 0;
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (area[i][j] >= height && !chk[i][j]) {
                    cnt++;
                    dfs(i, j);
                }
            }
        }
        if (cnt > maxCnt) maxCnt = cnt;
    }
    return maxCnt;
}

console.log(findMaxHeight());
